src/ipa/rpi/controller/rpi/sync.cpp

```diff
@@ -119,100 +144,198 @@ void Sync::initialise()
 	socket_ = -1;
 }
 
+/* No idea what this does*/
 void Sync::switchMode([[maybe_unused]] CameraMode const &cameraMode, [[maybe_unused]] Metadata *metadata)
 {
 	syncReady_ = false;
 	frameCount_ = 0;
 	readyCountdown_ = 0;
 }
 
+
+/* Most important part, algorithm*/
 void Sync::process([[maybe_unused]] StatisticsPtr &stats, Metadata *imageMetadata)
 {
 	SyncPayload payload;
-	SyncParams local {};
-	SyncStatus status {};
+	SyncParams local{};
+	SyncStatus status{};
 	imageMetadata->get("sync.params", local);
+	static int64_t lag = 0;
+	static bool initialise = false;
+
 
 	if (!frameDuration_) {
 		LOG(RPiSync, Error) << "Sync frame duration not set!";
 		return;
 	}
+	if (mode_ == Mode::Off) {
+		return;
+	}
 
 	if (mode_ == Mode::Server) {
+
+		if (!initialise){
+			initialise = true;
+			Sync::initialise_socket();
+		}
+
+		static ClockRecovery trendingClock(local.wallClock, local.sensorTimestamp, syncPeriod_, lineFitting_);
+		
+		static int64_t lastWallClock = local.wallClock;
+
+		int mul = (local.wallClock - lastWallClock - frameDuration_.get<std::micro>() * 0.5) / frameDuration_.get<std::micro>();
+		if (mul > 0) {
+			frameCount_ += mul;
+		}
+		lastWallClock = local.wallClock;
+
+		static int64_t syncTime = local.wallClock + (readyFrame_ - frameCount_) * int(frameDuration_.get<std::micro>());
+
 		if (!syncReady_ && !(readyFrame_ - frameCount_)) {
-			LOG(RPiSync, Info) << "Sync ready at frame " << frameCount_ << " ts " <<  payload.wallClock;
-			syncReady_ = true;
+			if (local.wallClock > syncTime - frameDuration_.get<std::micro>() * 0.5) {
+				LOG(RPiSync, Error) << "Sync ready is true";
+				syncReady_ = true;
+				lag = local.wallClock - syncTime;
+				if (local.wallClock > syncTime + frameDuration_.get<std::micro>() * 0.5) {
+					LOG(RPiSync, Warning) << "Frame has been lost, sync started with lag of: " << lag << " us";
+				} else {
+					LOG(RPiSync, Info) << "Sync started without lag";
+				}
+			}
+		} else if (!syncReady_) {
+			syncTime = local.wallClock + frameDuration_.get<std::micro>() * (readyFrame_ - frameCount_);
 		}
 
 		if (!(frameCount_ % syncPeriod_)) {
```

Comment:
Here too, `frameCount_` could jump just over the next multiple of `syncPeriod`. I wonder if we need another counter which counts up to `syncPeriod` (or just beyond), which we then reset to zero every time we send a packet?

Reply:
We ave that, it is called frames